### PR TITLE
Remove Content-Type auto-detection

### LINKED ISSUE #6092: Traefik 2.0 dropping Content-Type headers
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I set up the docker image referenced: https://github.com/sissbruecker/linkding

I added labels to expose it to the HTTPS entrypoint with proper domain, etc.

Navigate to site behind traefik and see console error that the stylesheet is being server wrong.

Open dev tools, run it again, and see the stylesheet is being returned text/plain

### What did you expect to see?
I expected to see the stylesheet being returned as text/css, like it does when I navigate to the direct IP of the docker container.

### What did you see instead?
I saw the stylesheet returned as text/plain, which will not load as valid CSS over HTTPS

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.7
Codename:     montdor
Go version:   go1.13.5
Built:        2019-12-09T17:51:21Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
providers:
  docker:
    endpoint: "unix:///var/run/docker.sock"
    exposedByDefault: false
    network: slenknet_default

entryPoints:
  websecure:
    address: ":443"
  metrics:
    address: ":9990"

http:
  middlewares:
    slenk-auth:
      basicAuth:
        users:
          - "######################"

certificatesResolvers:
  mytlschallenge:
    acme:
      email: my@email.com
      storage: "/letsencrypt/acme.json"
      tlsChallenge: {}

api:
  insecure: true
  dashboard: true
  debug: true

metrics:
  prometheus: 
    addEntryPointsLabels: true
    addServicesLabels: true
    entryPoint: metrics

accessLog: {}

```

```labels:
        - "traefik.enable=true"
        - "traefik.http.routers.linkding.rule=Host(`linkding.######.###`)"
        - "traefik.http.routers.linkding.tls=true"
        - "traefik.http.routers.linkding.entrypoints=websecure"
        - "traefik.http.routers.linkding.tls.certresolver=mytlschallenge"
        - "traefik.http.services.linkding.loadbalancer.server.port=9090"
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
(paste your output here)
```

### PR DESCRIPTION
**Pull request by juliens**
### What does this PR do?

This PR removes the auto-detection of the `Content-Type`.

### Motivation
When the backend doesn't send a `Content-Type` header, Traefik tried to auto-detect a `Content-Type`. For a CSS, Traefik will detect `text/plain` and browser will not interperet css if it's in `text/plain` and will interpret it if there is no `Content-Type`.

Fixes #6092 #4761 #3131

### More

- [x] Added/updated tests
- [ ] Added/updated documentation

Co-Authored-By: mpl <mathieu.lonjaret@gmail.com>

### CODE REVIEW

**Comment by dtomcej** [comment]
Obviously this is an issue created by upstream code, and may be subject to change in the future.

However, this design looks good to me!

Design LGTM!

**Comment by derekslenk** [comment]
Sorry if this is inappropriate, I am just very interested in this issue. Does the label v2 mean we will have to wait until 2.2? Is there an image published with this on it and/or is it possible to build it myself?

**Comment by mpl** [comment]
> Sorry if this is inappropriate, I am just very interested in this issue. Does the label v2 mean we will have to wait until 2.2?

No, as this is considered a bugfix, it should be merged to the 2.1 branch. So it will be in the next 2.1.x release.

> Is there an image published with this on it and/or is it possible to build it myself?

I think a couple of days after it is merged in 2.1, the traefik:v2.1 image should include that fix.
Otherwise, yes, you can build it yourself. Just pull the code from that pull request into your local repo, and run "make build-image". It should create the "containous/traefik:latest" image locally.

File: pkg/config/dynamic/middlewares.go [hunk 1]
```diff
@@ -35,6 +35,14 @@ type Middleware struct {
...
 	PassTLSClientCert *PassTLSClientCert `json:"passTLSClientCert,omitempty" toml:"passTLSClientCert,omitempty" yaml:"passTLSClientCert,omitempty"`
 	Retry             *Retry             `json:"retry,omitempty" toml:"retry,omitempty" yaml:"retry,omitempty"`
+	ContentType       *ContentType       `json:"contentType,omitempty" toml:"contentType,omitempty" yaml:"contentType,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+
+// ContentType holds the ContentType configuration.
+type ContentType struct {
+	AutoDetect bool `json:"autoDetect,omitempty" toml:"autoDetect,omitempty" yaml:"autoDetect,omitempty"`
```
Comment (mpl) [inline pkg/config/dynamic/middlewares.go]:
```suggestion
	// AutoDetect specifies whether to let the Content-Type header, if it hasn't
	// been set by the backend, be automatically set to a value derived from the
	// contents of the response. As a proxy, the default behaviour should be to leave
	// the header alone, regardless of what the backend did with it. Unfortunately, the
	// historic default was to always auto-detect and set the header if it was nil, and
	// it is going to be kept that way in order to support users currently relying on
	// it. This middleware exists to enable the correct behaviour until the default
	// one can be changed in a future version.
	AutoDetect bool `json:"autoDetect,omitempty" toml:"autoDetect,omitempty" yaml:"autoDetect,omitempty"`
```

File: docs/content/middlewares/contenttype.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+!!! info
+    
+    For compatibility reason, the default behaviour on a router (without this middleware) , is to auto-detect the `Content-Type`.
+
+## Configuration Options
+
+### `autoDetect`
+
+`autoDetect` specifies if we want to auto-detect `Content-Type`.
```
Comment (mpl) [inline docs/content/middlewares/contenttype.md]:
```suggestion
`autoDetect` specifies whether to let the `Content-Type` header, if it hasn't
been set by the backend, be automatically set to a value derived from the
contents of the response. As a proxy, the default behaviour should be to leave
the header alone, regardless of what the backend did with it. Unfortunately, the
historic default was to always auto-detect and set the header if it was nil, and
it is going to be kept that way in order to support users currently relying on
it. This middleware exists to enable the correct behaviour until at least the
default one can be changed in a future version.
```

File: docs/content/middlewares/contenttype.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+http:
+  middlewares:
+    autodetect:
+      contentType:
+        autoDetect: false
+```
+
+!!! info
+    
+    For compatibility reason, the default behaviour on a router (without this middleware) , is to auto-detect the `Content-Type`.
```
Comment (mpl) [inline docs/content/middlewares/contenttype.md]:
```suggestion
    For compatibility reason, the default behaviour on a router (without this middleware) , is to automatically set the `Content-Type` header (if it was unset), with a value derived from the contents of the response. Therefore, given the default value of the `autoDetect` option, simply enabling this middleware for a router potentially changes its behaviour.
```

**Pull request merged**